### ISSUE TITLE
Convolution operations are extremely slow on RTX 30 series GPU

### ISSUE THREAD

**Issue by P3n9W31** [pytorch/pytorch#47039]
## 🐛 Bug

I'm using PyTorch on an RTX 3090 GPU, and trying to train different types of networks (e.g., CNN, RNN, Transformer), but the training speed is really abnormal when training the CNN. To illustrate, another RTX 2080ti was used for the comparison:

(All the data in the table are in "s/batch", i.e, seconds per batch, and under the same network architecture(row).)

|                    | RTX 3090 | RTX 2080ti | CPU   |
| ------------------ | -------- | ---------- | ----- |
| CNN (TextCNN)      | **3.95** | 0.33       | 15.44 |
| RNN (LSTM)         | 0.1862   | 0.38       | 16.25 |
| Transformer (BERT) | 0.14     | 0.32       | 15.06 |

I tested it on two versions of pytorch: 

1. torch == 1.8.0.dev20201028+cu110 
2. torch == 1.7.0+cu110

The results are the same.

And, I tested it on: 

```python
torch.backends.cudnn.benchmark=True
```

and

```python
torch.backends.cudnn.benchmark=False
```

The results are still the same - **Convolution operations are extremely slow on RTX 3090** .

## To Reproduce

As previously stated.

## Expected behavior

Convolution operations on RTX 3090 should faster than (or close to) on RTX 2080ti.

## Environment

```
PyTorch version: 1.8.0.dev20201028+cu110
Is debug build: True
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 8.4.0-3ubuntu2) 8.4.0
Clang version: Could not collect
CMake version: version 3.18.2

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GeForce RTX 3090
Nvidia driver version: 455.23.05
cuDNN version: Probably one of the following:
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.8.0.dev20201028+cu110
[pip3] torchaudio==0.7.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.7.0
[pip3] torchvision==0.9.0.dev20201027+cu110
[conda] Could not collect
```

😕What could be the reason for such strange results only when performing a convolution?

**Comment by ptrblck** [comment]
The binaries use cudnn8.0.3, which doesn't ship with tuned heuristics for 3090 and cudnn8.0.5 will provide them.
Until then performance regressions on these devices are unfortunately expected.

**Comment by P3n9W31** [comment]
> The binaries use cudnn8.0.3, which doesn't ship with tuned heuristics for 3090 and cudnn8.0.5 will provide them.
> Until then performance regressions on these devices are unfortunately expected.

Very thanks for your help.😃
So the solution may be: I should compile PyTorch from source once when cuDNN 8.0.5 is available？

**Comment by ptrblck** [comment]
Yes, that should help once it's released.
We are currently verifying it internally.

**Comment by albanD** [comment]
Closing as this looks solved: current binaries are not tuned and reduced perfs are expected.

**Comment by P3n9W31** [comment]
Some new facts discovered.

The current 30-series GPUs don't perform well on convolution. But the defects will not show up at all times, but when doing some dimension-specific convolution.

Specifically, since I'm using CNN on a natural language task, the text sequences are of different lengths. To speed up the training, I started out using the dynamic padding method, i.e., padding each piece of text to the maximum length of the current batch, which resulted in each batch being of a different length. At this point, the training process does not take 3.95 seconds for each batch as I mentioned before, and some batches are extremely slow while others seem normal, so the training speed is not stable. (may looks like: 3.95s/batch, 0.3s/batch, 0.3s/batch, 3.95s/batch...)

And I thought the problem is: cuDNN will first choose the most suitable algorithm according to the convolution dimension before doing the calculation, so the process of finding the algorithm will be repeated many times when the length of each batch is different, which makes the calculation very time-consuming. (On the 2080ti, this process is done very well, and there should be a specially optimized heuristic algorithm.)

I tried to pad each batch to the same length (not using dynamic padding) and then measured the computation time again. I found that when the data is padded to length 512, the speed is normal (0.3558 s/batch, close to 2080ti), but when padded to length 256, it goes back to 3.95 s/batch again. (Convolution speed varies by dimension)

So, maybe, indeed, some optimizations are missing.

(Hope cuDNN 8.0.5 will be released soon.)

**Comment by pdima** [comment]
I can confirm much worse performance with 3090 comparing with 2080ti with a common CNN like efficientnet etc. using CUDA 11.1, cuDNN 8.0.4 and the source pytorch build from 3 Nov.

**Comment by irishev** [comment]
I've found that the numbers of channels and groups can affect the performance significantly, even though they are divisible by 8. The binaries are needed to be tuned.

**Comment by megamuzzy** [comment]
[cuDNN 8.0.5](https://docs.nvidia.com/deeplearning/cudnn/release-notes/rel_8.html#rel_8) has been released. I have updated it yesterday, but I haven't noticed convolution performance improvement...

**Comment by ptrblck** [comment]
cudnn8.0.5 should improve the performance on the RTX 3090. Which device are you using and which conv setup?

**Comment by megamuzzy** [comment]
I am using two 3-layer conv models (conv1d and conv2d) on RTX 3090:

Conv1dModel(
  (conv): Sequential(
    (0): Conv1d(24, 256, kernel_size=(12,), stride=(6,), groups=4)
    (1): ReLU()
    (2): Conv1d(256, 256, kernel_size=(6,), stride=(3,), padding=(2,), groups=4)
    (3): ReLU()
    (4): Conv1d(256, 256, kernel_size=(3,), stride=(1,), padding=(1,), groups=4)
    (5): ReLU()
  )
)
Conv2dModel(
  (conv): Sequential(
    (0): Conv2d(8, 32, kernel_size=(8, 8), stride=(4, 4))
    (1): ReLU()
    (2): Conv2d(32, 64, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1))
    (3): ReLU()
    (4): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (5): ReLU()
  )
)

The performance on RTX 3090 is slightly worse than on 1080ti (my previous setup)

**Comment by ptrblck** [comment]
Thanks for the model. I don't have a 1080Ti ready, so used a P100 as a proxy and see a speedup for the 3090 and overall better performance:
```
3090
8.0.4.30
Conv1d model, benchmark=False, 0.472ms/iter
Conv1d model, benchmark=True, 0.495ms/iter
Conv2d model, benchmark=False, 1.630ms/iter
Conv2d model, benchmark=True, 1.467ms/iter

8.0.5.35
Conv1d model, benchmark=False, 0.248ms/iter
Conv1d model, benchmark=True, 0.248ms/iter
Conv2d model, benchmark=False, 1.390ms/iter
Conv2d model, benchmark=True, 1.301ms/iter

Tesla P100-SXM2-16GB
8.0.4.30
Conv1d model, benchmark=False, 0.950ms/iter
Conv1d model, benchmark=True, 0.885ms/iter
Conv2d model, benchmark=False, 2.681ms/iter
Conv2d model, benchmark=True, 2.701ms/iter

8.0.5.35
Conv1d model, benchmark=False, 0.962ms/iter
Conv1d model, benchmark=True, 0.910ms/iter
Conv2d model, benchmark=False, 2.672ms/iter
Conv2d model, benchmark=True, 2.695ms/iter
```

Code:
```python
import torch
import torch.nn as nn
import time

def profile(model, x, benchmark, nb_iters):
    torch.backends.cudnn.benchmark = benchmark

    # warmup
    for _ in range(10):
        out = model(x)

    torch.cuda.synchronize()
    t0 = time.time()
    for _ in range(nb_iters):
        out = model(x)
    torch.cuda.synchronize()
    t1 = time.time()

    return (t1 - t0) / nb_iters


model1 = nn.Sequential(
    nn.Conv1d(24, 256, kernel_size=(12,), stride=(6,), groups=4),
    nn.ReLU(),
    nn.Conv1d(256, 256, kernel_size=(6,), stride=(3,), padding=(2,), groups=4),
    nn.ReLU(),
    nn.Conv1d(256, 256, kernel_size=(3,), stride=(1,), padding=(1,), groups=4),
    nn.ReLU(),
).cuda()

x = torch.randn(64, 24, 224, device='cuda')
print(torch.cuda.get_device_name(0))
time0 = profile(model1, x, benchmark=False, nb_iters=100)
print('Conv1d model, benchmark=False, {:.3f}ms/iter'.format(time0*1000))
time1 = profile(model1, x, benchmark=True, nb_iters=100)
print('Conv1d model, benchmark=True, {:.3f}ms/iter'.format(time1*1000))

model2 = nn.Sequential(
    nn.Conv2d(8, 32, kernel_size=(8, 8), stride=(4, 4)),
    nn.ReLU(),
    nn.Conv2d(32, 64, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1)),
    nn.ReLU(),
    nn.Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1)),
    nn.ReLU()
).cuda()

x = torch.randn(64, 8, 224, 224, device='cuda')
time0 = profile(model2, x, benchmark=False, nb_iters=100)
print('Conv2d model, benchmark=False, {:.3f}ms/iter'.format(time0*1000))
time1 = profile(model2, x, benchmark=True, nb_iters=100)
print('Conv2d model, benchmark=True, {:.3f}ms/iter'.format(time1*1000))
```

**Comment by megamuzzy** [comment]
This is what I got on my setup :(

GeForce RTX 3090
Conv1d model, benchmark=False, 0.867ms/iter
Conv1d model, benchmark=True, 0.869ms/iter
Conv2d model, benchmark=False, 2.688ms/iter
Conv2d model, benchmark=True, 2.591ms/iter

**Comment by pdima** [comment]
I have about 4x performance improvement after upgrading cudnn from 8.0.4 to 8.0.5.

**Comment by megamuzzy** [comment]
I have two python installations in my system. Ubuntu preinstalled python 3.6.9 and conda env with python 3.7.9 
And `torch.backends.cudnn.version() ` reports different versions of cudnn:
```
GeForce RTX 3090
sys.version= 3.6.9 (default, Oct  8 2020, 12:12:24) 
[GCC 8.4.0]
torch.backends.cudnn.version()= 8004
torch.__version__= 1.7.0+cu110
Conv1d model, benchmark=False, 0.376ms/iter
Conv1d model, benchmark=True, 0.399ms/iter
Conv2d model, benchmark=False, 1.495ms/iter
Conv2d model, benchmark=True, 1.516ms/iter

GeForce RTX 3090
sys.version= 3.7.9 (default, Aug 31 2020, 12:42:55) 
[GCC 7.3.0]
torch.backends.cudnn.version()= 8003
torch.__version__= 1.7.0
Conv1d model, benchmark=False, 0.861ms/iter
Conv1d model, benchmark=True, 0.861ms/iter
Conv2d model, benchmark=False, 2.574ms/iter
Conv2d model, benchmark=True, 2.647ms/iter

```
 I have reinstalled `libcudnn8_8.0.5.39-1+cuda11.0_amd64.deb` several tmes, but it still reports 8003 and 8004.
Please tell me how to properly reinstall cudnn? Thanks!

**Comment by P3n9W31** [comment]
I tested the 3090's performance using code provided by ptrblck before upgrading the cuDNN. And I got this:

```
GeForce RTX 3090
torch.backends.cudnn.version()= 8004
torch.__version__= '1.8.0.dev20201111+cu110'
Conv1d model, benchmark=False, 0.249ms/iter
Conv1d model, benchmark=True, 0.248ms/iter
Conv2d model, benchmark=False, 1.531ms/iter
Conv2d model, benchmark=True, 1.514ms/iter
```

It seems that the reported speed is similar to that obtained by ptrblck. But when I rerun the TextCNN, I still got a super slow speed. 

It looks like this code doesn't reproduce my problem very well, I will try to give a minimal implementation.

And before that, I really need some help with the upgrading. Which version of PyTorch should I clone when I am trying to install PyTorch from source. I have tried master and 'nightly' branch, both of them throw a warning - “capability sm_86 is not compatible with the current PyTorch installation”.

**Comment by hhcs9527** [comment]
you may need to export TORCH_CUDA_ARCH_LIST="8.0" .

**Comment by tgalery** [comment]
Hi there, I've been experiencing a similar issue with an RTX 3060 TI. I'm using ubuntu 20.04 on wsl. I've installed cuda 11.2 and cudnn 8.2. I've built pytorch from scratch to see if it would alleviate the problem and essentially I get the following results:


```
NVIDIA GeForce RTX 3060 Ti
Conv1d model, benchmark=False, 11.444ms/iter
Conv1d model, benchmark=True, 12.046ms/iter
Conv2d model, benchmark=False, 4.893ms/iter
Conv2d model, benchmark=True, 4.359ms/iter
```

This seems dog slow, no ?

**Comment by ML-Mr-J** [comment]
> I have about 4x performance improvement after upgrading cudnn from 8.0.4 to 8.0.5.

so which way did you choose compiling pytorch or pip 1.8.x with cudnn111 from official website of torch.

**Comment by QuanticDisaster** [comment]
```
NVIDIA GeForce RTX 3060
Conv1d model, benchmark=False, 0.309ms/iter
Conv1d model, benchmark=True, 0.309ms/iter
Conv2d model, benchmark=False, 3.311ms/iter
Conv2d model, benchmark=True, 3.291ms/iter
```

Here are my results, while the conv1d seems close to what ptrblck got, the conv2d model seems slower than some of the precedent results, should i worry about it ?

I am working with windows 10 on an anaconda virtual environnement with
pytorch                   1.8.0           py3.8_cuda11.1_cudnn8_0    pytorch
cudatoolkit               11.1.1
torch.backends.cudnn.version() = 8005

**Comment by shengcheng** [comment]
Here is my results: 
py3.6_cuda11.0.221_cudnn8.0.5_0 + cuda 11.0 on V100 and A100:
`Tesla V100-PCIE-16GB
Conv1d model, benchmark=False, 0.675ms/iter
Conv1d model, benchmark=True, 0.616ms/iter
Conv2d model, benchmark=False, 6.487ms/iter
Conv2d model, benchmark=True, 6.483ms/iter`

`A100-PCIE-40GB
Conv1d model, benchmark=False, 0.825ms/iter
Conv1d model, benchmark=True, 0.457ms/iter
Conv2d model, benchmark=False, 1.801ms/iter
Conv2d model, benchmark=True, 1.457ms/iter`

`py3.6_cuda9.0.176_cudnn7.5.1_0 + cuda9.0 on V100:
Tesla V100-PCIE-16GB
Conv1d model, benchmark=False, 0.457ms/iter
Conv1d model, benchmark=True, 0.411ms/iter
Conv2d model, benchmark=False, 1.744ms/iter
Conv2d model, benchmark=True, 1.440ms/iter`

So, for the same GPU(V100), the performance is worse for newest pytorch/cuda/cudnn version. For A100, it does not have the performance gain as it should have.

**Comment by timmyvg** [comment]
I have the same problem with RTX A6000. I also notice my CNNs model run supper slower than RTXA3090. 
torch.backends.cudnn.version() 
RTX A6000 (48GB)
Pytorch 1.8, cuda 11.2, torch.backends.cudnn.version()  = 8005
Self CPU time total: 800.873ms 
Self CUDA time total: 807.593ms 

RTX 3090 (24 GB) 
Pytorch 1.8, cuda 11.3, torch.backends.cudnn.version()  = 8005
Self CPU time total: 268.730ms 
CUDA time total: 290.090ms 

RTX 2080Ti (12GB)
Pytorch 1.8, Cuda 10.2, torch.backends.cudnn.version()  = 7605
Self CPU time total: 509.958ms 
Self CUDA time total: 524.578ms 

I hope to see some solutions.

**Comment by andrecavalcante** [comment]
Hi, I am also getting slow results comparing 3090 and the 1080ti. I have built PyTorch and torchvision from source, CUDA 11.5 and cudnn 8.3.0.98, nvidia driver 925. Are these new versions optimized for the 3090? My results are below. Thanks everybody!

NVIDIA GeForce RTX 3090
Conv1d model, benchmark=False, 0.353ms/iter
Conv1d model, benchmark=True, 0.353ms/iter
Conv2d model, benchmark=False, 1.724ms/iter
Conv2d model, benchmark=True, 1.666ms/iter

NVIDIA GeForce GTX 1080 Ti
Conv1d model, benchmark=False, 0.274ms/iter
Conv1d model, benchmark=True, 0.274ms/iter
Conv2d model, benchmark=False, 2.127ms/iter
Conv2d model, benchmark=True, 2.067ms/iter

**Comment by FrancescoSaverioZuppichini** [comment]
same here

**Comment by Brilhador** [comment]
I have the same problem running 3090 vs 3060. Does anyone have a solution?

**Comment by testaibot** [comment]
I have the same problem running 3070ti Laptop vs 2080 Desktop with 1.27it/s vs 7.95it/s respectively

**Comment by taalhaataahir0102** [comment]
Same Cudnn 9.6.0
3050 Ti Laptop is slower
